pinot-core/src/main/java/org/apache/pinot/core/segment/processing/framework/SegmentMapper.java

```diff
@@ -0,0 +1,135 @@
+import org.apache.pinot.core.segment.processing.utils.SegmentProcessorUtils;
+import org.apache.pinot.spi.data.readers.GenericRow;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Mapper phase of the SegmentProcessorFramework.
+ * Reads the input segment and creates partitioned avro data files
+ * Performs:
+ * - record transformations
+ * - partitioning
+ * - partition filtering
+ */
+public class SegmentMapper {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentMapper.class);
+  private final File _inputSegment;
+  private final File _mapperOutputDir;
+
+  private final String _mapperId;
+  private final Schema _avroSchema;
+  private final RecordTransformer _recordTransformer;
+  private final RecordFilter _recordFilter;
+  private final Partitioner _partitioner;
+  private final Map<String, DataFileWriter<GenericData.Record>> _partitionToDataFileWriterMap = new HashMap<>();
+
+  public SegmentMapper(String mapperId, File inputSegment, SegmentMapperConfig mapperConfig, File mapperOutputDir) {
+    _inputSegment = inputSegment;
+    _mapperOutputDir = mapperOutputDir;
+
+    _mapperId = mapperId;
+    _avroSchema = SegmentProcessorUtils.convertPinotSchemaToAvroSchema(mapperConfig.getPinotSchema());
+    _recordTransformer = RecordTransformerFactory.getRecordTransformer(mapperConfig.getRecordTransformerConfig());
+    _recordFilter = RecordFilterFactory.getRecordFilter(mapperConfig.getRecordFilterConfig());
+    _partitioner = PartitionerFactory.getPartitioner(mapperConfig.getPartitioningConfig());
+    LOGGER.info(
+        "Initialized mapper with id: {}, input segment: {}, output dir: {}, recordTransformer: {}, recordFilter: {}, partitioner: {}",
+        _mapperId, _inputSegment, _mapperOutputDir, _recordTransformer.getClass(), _recordFilter.getClass(),
+        _partitioner.getClass());
+  }
+
+  /**
+   * Reads the input segment and generates partitioned avro data files into the mapper output directory
+   * Records for each partition are put into a directory of its own withing the mapper output directory, identified by the partition name
+   */
+  public void map()
+      throws Exception {
+
+    PinotSegmentRecordReader segmentRecordReader = new PinotSegmentRecordReader(_inputSegment);
+    GenericRow reusableRow = new GenericRow();
+    GenericData.Record reusableRecord = new GenericData.Record(_avroSchema);
+
+    while (segmentRecordReader.hasNext()) {
+      reusableRow = segmentRecordReader.next(reusableRow);
+
+      // Record transformation
+      reusableRow = _recordTransformer.transformRecord(reusableRow);
+
+      // Record filtering
+      if (_recordFilter.filter(reusableRow)) {
+        continue;
+      }
+
+      // Partitioning
+      String partition = _partitioner.getPartition(reusableRow);
+
+      // Create writer for the partition, if not exists
+      if (!_partitionToDataFileWriterMap.containsKey(partition)) {
```

Comment:
Use `_partitionToDataFileWriterMap.get(partition)` and check if the value is null to save one map lookup